### PR TITLE
Enhance AI tools sampling logic for robust handling of large fields

### PR DESCRIPTION
**Pull request by sgrebnov**
# 🗣 Description

Current sampling logic can return very large data that could not be handled by AI models, plus expensive. PR adds logic to truncate large text fields

## 🔨 Related Issues

Fixes NSQL bug: string too long. Expected a string with maximum length 1048576, but got a string with length 1065155 instead. https://github.com/spiceai/spiceai/issues/3946

## 🤔 Concerns

We might need to make size parameters configurable

### CODE REVIEW

File: crates/runtime/src/tools/builtin/sample/tool.rs [hunk 1]
```diff
@@ -91,7 +92,11 @@ impl SpiceModelTool for SampleDataTool {
         let span: Span = tracing::span!(target: "task_history", tracing::Level::INFO, "tool_use::sample_data", tool = self.name(), input = format!("{params}"), sample_method = self.method.name());
 
         async {
-            let batch = params.sample(rt.datafusion()).await?;
+            let mut batch = params.sample(rt.datafusion()).await?;
+
+            // truncate large text fields
+            batch = truncate_record_batch_data(&batch, 512)?;
```
Comment (sgrebnov) [inline crates/runtime/src/tools/builtin/sample/tool.rs]: @Jeadie - what would be the recommended default max size?

File: crates/arrow_tools/src/record_batch.rs [hunk 2]
```diff
@@ -138,6 +138,119 @@ pub fn to_primitive_type_list(
...
+        }
+        _ => Ok(column),
+    }
+}
+
+fn trancate_str(str: Option<&str>, max_characters: usize) -> Option<&str> {
+    match str {
+        Some(value) => {
+            if value.len() > max_characters {
+                // Some(String::from(&value[..max_characters]))
```
Comment (Jeadie) [inline crates/arrow_tools/src/record_batch.rs]:
```suggestion

```

File: crates/arrow_tools/src/record_batch.rs [hunk 3]
```diff
@@ -138,6 +138,119 @@ pub fn to_primitive_type_list(
     Err(ArrowError::CastError("Invalid column type".into()))
 }
 
+/// Recursively truncates the data in a record batch to the specified maximum number of characters.
+/// The truncation is applied to `Utf8` and `Utf8View` data.
+/// # Errors
+///
+/// This function will return an error if arrow conversion fails.
```
Comment (phillipleblanc) [inline crates/arrow_tools/src/record_batch.rs]:
```suggestion
/// Recursively truncates the data in a record batch to the specified maximum number of characters.
/// The truncation is applied to `Utf8` and `Utf8View` data.
///
/// # Errors
///
/// This function will return an error if arrow conversion fails.
```

File: crates/arrow_tools/src/record_batch.rs [hunk 4]
```diff
@@ -138,6 +138,119 @@ pub fn to_primitive_type_list(
...
+pub fn truncate_record_batch_data(
```
Comment (phillipleblanc) [inline crates/arrow_tools/src/record_batch.rs]: `truncate_string_columns`?

File: crates/arrow_tools/src/record_batch.rs [hunk 5]
```diff
@@ -138,6 +138,119 @@ pub fn to_primitive_type_list(
...
+        .columns()
+        .iter()
+        .zip(schema.fields())
+        .map(|(column, field)| truncate_column_data(Arc::clone(column), field, max_characters))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    RecordBatch::try_new(schema, columns)
+}
+
+fn truncate_column_data(
```
Comment (phillipleblanc) [inline crates/arrow_tools/src/record_batch.rs:162]:
```suggestion
fn truncate_string_column_data(
```

File: crates/arrow_tools/src/record_batch.rs (hunk 4 above)
Comment (sgrebnov) [reply crates/arrow_tools/src/record_batch.rs]: Like it - thank you

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**